### ISSUE TITLE
1.4.7 - Access Control Documentation

### ISSUE THREAD

**Issue by EnigmaRosa** [OWASP/ASVS#2065]
Note: This is referred to as 4.1.6 in #2033 but it was determined that it belonged in v1 instead.

Proposing a documentation requirement for the access control system design, as it is currently missing from v1.

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **1.4.7** | [Added] Verify documentation and definition of all user, resource, action, and environmental attributes and access control policies. | ✓ | ✓ | ✓ | 284|

**Comment by elarlang** [comment]
This is a requirement we need to have. I wrote related comment to other issue (https://github.com/OWASP/ASVS/issues/2058#issuecomment-2330627854) - we just need to analyze, that the documentation covers the need for all V4 implementation requirements.

Comment from Josh (https://github.com/OWASP/ASVS/issues/2033#issuecomment-2324157184):

> However, I think the wording needs to be simpler/clearer.

@tghosth - any direction or proposal what should be improved?

**Comment by elarlang** [comment]
When it is PR time, note that I moved away added 1.4.6, so the requirement can have a number 1.4.6 with the label `[ADDED]` (I'll check and take care of it when it is PR-ready)

**Comment by jmanico** [comment]
Note, documenting access control policy is so fundamental to good access control that I would like to keep it in the access control section. I do not want developers to have to hop to a variety of different sections to get access control right.

**Comment by elarlang** [comment]
If we have structure to keep all documentation requirements in V1, then we do that. The point is to point out questions for the analysis step and to provide a separate set of requirements. If this requirement should be located somewhere else, most likely it will be missed/skipped at that step.

What we going to do with current V1 content is up to separate discussion https://github.com/OWASP/ASVS/issues/1063 / https://github.com/OWASP/ASVS/discussions/1831

**Comment by jmanico** [comment]
My point is, if developers are building access control then the most important thing - since its business logic - is to implement what was documented and to maintain that documentation. I worry this will be missed if it's only in v1.

**Comment by tghosth** [comment]
I think my challenge with this requirement is that I am not sure that I understand exactly what it is talking about:

> _Verify documentation and definition of_

That seems clear enough

> _all user, resource, action,_ 

Are we talking about the different dimensions where access control has to be enforced? 

For example:
* Is the user allowed to interact with this particular entity type (e.g. orders in an eCommerce system)
* Is the user allowed to perform this particular action on this entity type (e.g. can they create an order or only view an order)
* Is the user allowed to interact with this particular instance of this entity type (e.g. can they only see or modify orders for their department but not other departments)
* Is the user allowed to interact with these particular fields of this particular entity type (e.g. they can view order details but they are not allowed to see the standard discount amount)

(If I have understood this correctly, does the 4th bullet this solve Problem 1 in [this comment](https://github.com/OWASP/ASVS/issues/934#issuecomment-2325740099) @elarlang?)

> _and environmental attributes_ 

We talk about environmental in the context of where the user is located somewhere else, is that what we are talking about here?

> _and access control policies._

Not sure what we are adding here.

**Comment by EnigmaRosa** [comment]
@tghosth The term "attributes" (whether they are user, resource, action, or environmental) references attribute based access control (ABAC) where attributes describe the user (user ID, group membership, etc.), resource (file type, creation date, some other metadata, etc), action (read, write, edit), and environment (contextual information, this can even include the type of encryption used). Meanwhile, it is the policy that actually sets the rule (i.e. users in the "admins" group can read/write/edit/delete all folders). Would it be helpful if I clarified this in the requirement?

**Comment by EnigmaRosa** [comment]
Makes sense. How about something along the lines of "Verify the documentation and definition of the application's access control system, including access control policies and any relevant information that is used to determine access"

**Comment by elarlang** [comment]
Few notes for last comments:
* "Should" is topic for recommendations and rises the question, is the text for a requirement or for a recommendation
* Although "clearly" is one of my "parasite words", I would not use it in the requirement text.
* `/` must be "and" or "or"

**Comment by jmanico** [comment]
That is helpful, @elarlang 

How about:

_Verify that access control documentation defines the rules for access control decision-making, specifying user attributes, subject attributes, resource attributes, and relevant environmental factors involved in the process._

**Comment by tghosth** [comment]
I think user and subject are interchangable. Whilst user is more understandable I think subject is more correct.

I would suggest:

> Verify that access control documentation defines the rules for access control decision-making, specifying user and subject attributes, resource attributes, and relevant environmental factors involved in the process.

I think some explanation in the section heading would still be needed.

Any further comments?

**Comment by elarlang** [comment]
Let's get it in. If someone has ideas for improvements we can do it later.

Would @EnigmaRosa like to handle the PR or we do?

**Comment by EnigmaRosa** [comment]
I’ll handle the PR

**Comment by elarlang** [comment]
Added to **V1.4 Access Control Architecture** via #2123

| # | Description | L1 | L2 | L3 | CWE |
| :---: | :--- | :---: | :---: | :---: | :---: |
| **1.4.7** | [ADDED] Verify that access control documentation defines the rules for access control decision-making, specifying user and subject attributes, resource attributes, and relevant environmental factors involved in the process. | ✓ | ✓ | ✓ | |
